### ISSUE TITLE
Failure to warn of possible numeric comparison with an optional function parameter

### ISSUE THREAD

**Issue by WhyEvenTry** [Microsoft/TypeScript#12788]
**TypeScript Version:**  2.1.4

I had an issue with Flow that it warned for no reason because I had an "if" to guard against `undefined` (https://github.com/facebook/flow/issues/2984) - so I tried TypeScript. TypeScript didn't warn - **but** it also didn't warn when I removed the ifs!

Here is the code I used to check with TypeScript 2.1.4. Strict null checks are "true", see tsconfig.json below. I only get a warning about the unused `const test`.

**Code**

```ts
'use strict';

const imap: {search(criteria: string[], cb: Function): void} = {
    search: (criteria, cb) => {}
};

const search = (from?: number, to?: number): Promise<number[]> => {
    return new Promise((resolve, reject) => {
        imap.search(
            ['ALL'],
            (err: Error, uids: number[]) => {

                // NONE of this is safe, "from" could be undefined:
                const test: boolean = uids[0] >= from;
                uids = uids.filter((uid: number) => uid >= from);
                uids = uids.filter(function (uid: number) {
                    return uid >= from;
                });

            }
        )
    });
};
```

tsconfig.json:

```json
{
    "compilerOptions": {
        "strictNullChecks": true,
        "alwaysStrict": true,
        "module": "commonjs",
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "newLine": "lf",
        "noLib": false,
        "noUnusedLocals": true,
        "removeComments": true,
        "preserveConstEnums": true,
        "sourceMap": true,
        "lib": ["ES2017"],
        "target": "ESNext"
    },
    "include": [
        "lib/*"
    ],
    "exclude": [
        "node_modules",
        "**/*.spec.ts"
    ]
}
```

**Expected behavior:**

Error: `from`, used in a numeric comparison, could be undefined.

**Actual behavior:**

No warnings or errors other than the one about the unused variable `test`.

**Comment by RyanCavanaugh** [comment]
We consider comparing a `number` to `undefined` using a relational operator to be a safe operation. It's correct and useful to write code like this:
```ts
var x: number | undefined = ...;
if (x === 10) { // Safe: Could be a comparison between number and undefined, but OK
  // ... do something
}
```
and it's not clear why `>` ought to behave any differently since it's simply checking against range of numbers rather than a single number. It would be cumbersome and redundant to have to write `if ((x !== undefined) && (x > 0))` since the second comparison always returns `false` when the first one does anyway.

**Comment by WhyEvenTry** [comment]
The reason Flow correctly complains (although it over-complains) is that comparisons of a number against something that is not a number is useless.

**Comment by RyanCavanaugh** [comment]
> comparisons of a number against something that is not a number is useless.

Comparing `undefined` against a `number` using a relational operator always produces `false`, and it's entirely meaningful *and* idiomatic to use `===` *or* `>` to test for numbers to be present and equal or greater to some value.

Consider this code:
```ts
var x: number | undefined = ...;
if (x === 0) {
```
By your logic, we have to disallow this code, because we've compared a `number` to `undefined`, which is "useless"

So instead the author has to write
```ts
if ((x !== undefined) && (x === 0)) {
```
*which means the same thing*. As a code reviewer I would ask someone to remove the left condition. A type system which makes you write duplicative wrong-looking code isn't possibly helping you.

If you want to contend that `===` and `>` should behave differently, feel free, but I think that's a hard case to make.

**Comment by WhyEvenTry** [comment]
I#m not saying anything about "your code".  Comparing against something that is not a number may not affect your code. However, making the assumption that comparing something that is not a number against a number does not matter and is fine just because yo managed to find an example (of a single line of code...) where it actually really doesn't matter is *[self-censored word]*.

> So instead the author has to write

**NO!!!!** That would indeed be most useless.

Please read what I wrote.

**Comment by RyanCavanaugh** [comment]
Please don't argue at me about reading what you wrote when you edited your comment after-the-fact to add several new paragraphs of text.

**Comment by WhyEvenTry** [comment]
I always do that. It's the way I think. Alternatively, I could add 10 comments instead of one... would that be better? I can't help it, my ideas ALWAYS come in pieces.

**Comment by RyanCavanaugh** [comment]
Your options in that case are a) wait for your thoughts to come in before posting or b) don't complain about people not reading comments that haven't been written yet.

**Comment by RyanCavanaugh** [comment]
> I suggest you just ignore this ticket. No offense, there is no reason to be involved in every single issue

Responding to GitHub issues (among other things) is literally my job. Microsoft gives me reasons to be involved in the form of paychecks. If you'd like the TS team to ignore this ticket, why did you file it?

**Comment by WhyEvenTry** [comment]
..alternatively (to what I wrote in my scenario explanation five comments above): The situation could be an error that should not happen and makes no sense to continue as if nothing happens (which I already wrote) - but it could also be as you wrote, meaning it could be completely fine.

But in either case it would need to be made explicit, because it isn't obvious that continuing and ignoring that there was a nonsensical comparison is okay.

>  why did you file it?

Because see issue description. There is an issue.

**Comment by WhyEvenTry** [comment]
I admit I'm occasionally abrasive. Part of it is I don't really give a fuck about what you do with the ticket. Not my problem, I told you there is one, my job's done.

**Comment by WhyEvenTry** [comment]
I admit I'm occasionally abrasive. Part of it is I don't really give a fuck about what you do with the ticket. Not my problem, I told you there is one, my job's done. It also amuses me when people dismiss facts because they dislike my behavior, as if my behavior somehow changes the nature of the problem, magically.

If you "Intent" it to be that way such stupid behavior is not MY problem, I use the competition - which DOES complain.
